grove/amplification/grover.py

```diff
@@ -55,69 +57,72 @@ def _compute_grover_oracle_matrix(bitstring_map):
-        """Contructs an instance of Grover's Algorithm, using initialized values.
+    def _construct_grover_circuit(self) -> None:
+        """
+        Constructs an instance of Grover's Algorithm, using initialized values.
 
         :return: None
-        :rtype: NoneType
         """
-        oracle = pq.Program()
+        oracle = Program()
         oracle_name = "GROVER_ORACLE"
         oracle.defgate(oracle_name, self.unitary_function_mapping)
         oracle.inst(tuple([oracle_name] + self.qubits))
         self.grover_circuit = self.oracle_grover(oracle, self.qubits)
 
-    def _init_attr(self, bitstring_map):
-        """Initializes an instance of Grover's Algorithm given a bitstring_map.
+    def _init_attr(self, bitstring_map: Dict[str, int]) -> None:
+        """
+        Initializes an instance of Grover's Algorithm given a bitstring_map.
 
         :param bitstring_map: dict with string keys corresponding to bitstrings, and integer
          values corresponding to the desired phase on the output state.
         :type bitstring_map: Dict[String, Int]
         :return: None
-        :rtype: NoneType
         """
         self.bit_map = bitstring_map
         self.unitary_function_mapping = self._compute_grover_oracle_matrix(bitstring_map)
         self.n_qubits = self.unitary_function_mapping.shape[0]
         self.qubits = list(range(int(np.log2(self.n_qubits))))
         self._construct_grover_circuit()
 
-    def find_bitstring(self, cxn, bitstring_map):
+    def find_bitstring(self, qc: QuantumComputer, bitstring_map: Dict[str, int]) -> str:
         """
         Runs Grover's Algorithm to find the bitstring that is designated by ``bistring_map``.
 
         In particular, this will prepare an initial state in the uniform superposition over all bit-
         strings, an then use Grover's Algorithm to pick out the desired bitstring.
 
-        :param QVMConnection cxn: the connection to the Rigetti cloud to run pyQuil programs.
+        :param qc: the connection to the Rigetti cloud to run pyQuil programs.
         :param bitstring_map: a mapping from bitstrings to the phases that the oracle should impart
             on them. If the oracle should "look" for a bitstring, it should have a ``-1``, otherwise
             it should have a ``1``.
-        :type bitstring_map: Dict[String, Int]
         :return: Returns the bitstring resulting from measurement after Grover's Algorithm.
-        :rtype: str
         """
 
         self._init_attr(bitstring_map)
-        sampled_bitstring = cxn.run_and_measure(self.grover_circuit, self.qubits)[0]
-        return "".join([str(bit) for bit in sampled_bitstring])
+
+        ro = self.grover_circuit.declare('ro', 'BIT', len(self.qubits))
+        self.grover_circuit += [MEASURE(qubit, ro[idx]) for idx, qubit in enumerate(self.qubits)]
+        self.grover_circuit.wrap_in_numshots_loop(1)
+        executable = qc.compiler.native_quil_to_executable(self.grover_circuit)
```

Comment:
is there a reason why we can't do `executable = qc.compile(self.grover_circuit)`

Reply:
👍 just making sure there wasn't some problem with `qc.compile()` :) :)

Reply:
Now `qc.run(executable)` gives me a type exception message in PyCharm. It expects a `BinaryExecutableResponse` or a `PyQuilExecutableResponse` but gets a `Message` instead. Sure enough, [`QuantumComputer.compile`](https://github.com/rigetti/pyquil/blob/master/pyquil/api/_quantum_computer.py#L243) has `Message` as a return type. 

Does this change your mind about replacing `qc.compiler.native_quil_to_executable` with `qc.compile`, @mpharrigan?